### ISSUE TITLE
Drop `ApolloClient.query()` in next major

### ISSUE THREAD

**Issue by PowerKiKi** [apollographql/apollo-feature-requests#423]
I would like to suggest dropping  `ApolloClient.query()` entirely in the next major version. And instead all queries would have to use `watchQuery()` (which could potentially be renamed...).

The first reasons for this is to **reduce the API surface that has to be learned**. Because `watchQuery()` is pretty much a mandatory thing to learn if we want to leverage Apollo cache system, that means that pretty much everybody is already familiar with it (and its concept of emitting multiple results in time). As I see it, `query()`  is a less-flexible subset of of `watchQuery()`. And the sooner you understand that Apollo can, and should, emit multiple results over time, the better your overall DX will be.

Of course as it would reduce the API surface, it would simplify future maintenance of the library quite a bit. Eg, we wouldn't need `FetchPolicy` vs `WatchQueryFetchPolicy`. It would also most likely (slightly) reduce the bundle size.

This change would go especially well with the [migration to rxjs](https://github.com/apollographql/apollo-feature-requests/issues/375), since we could reproduce the original behavior of `query()` with something like:

```ts
apollo.watchQuery({query: ...})
   .pipe(first()) // Only the first result, like the old behavior of `query()`
   .subscribe();
```

It is not required to migrate to rxjs to make this breaking change, but rxjs syntax makes this much more reasonable.

I realize this will be a major change of habits, but that's what major versions are for, and the introduction of rxjs makes it a good time to do it.

**Comment by jerelmiller** [comment]
@PowerKiKi that is an interesting idea! 

I can understand why from an Angular perspective this might be beneficial, but I hesitate to remove it completely. Interestingly enough, I don't think we recommend `client.query()` enough on the React side where it would be easier for our users to use in certain situations. We tend to have some overuse of [`useLazyQuery`](https://www.apollographql.com/docs/react/api/react/hooks#uselazyquery) where a `client.query()` would be much more suitable.

Definitely something we will discuss as a team though. Appreciate the suggestion!

**Comment by jerelmiller** [comment]
At this time, we aren't planning to pursue this change. We are now recommending `client.query` in place of `useLazyQuery` for certain use cases in Apollo Client apps, so `client.query` still plays a big role.

That said, if you're keen on avoiding it, feel free to avoid exposing it in Apollo Angular 🙂. Thanks for the request!

**Comment by PowerKiKi** [comment]
Thanks for the feedback. Would you mind explaining in what kind of use cases you would recommend `client.query` nowadays ? Is it only for cases where we only need a single result and we want to specifically ignore all future cache updates ? I find those cases are quite rare...  Or is there something I am not aware of ?

**Comment by jerelmiller** [comment]
>  Is it only for cases where we only need a single result and we want to specifically ignore all future cache updates ?

Yep! I would agree the use cases are quite rare, but it is still useful for a good chunk of our users. For example, if you want to perform an imperative navigation after fetching, this might be a more suitable API since the component performing the fetch might otherwise get unmounted anyways.

Where we saw a lot of misuse in v3 was specifically with the `useLazyQuery` hook that we provided. Its a great hook to use when you want to execute the query in response to user interaction, but there were plenty of times we saw that users would consume the value from the promise, but ignore the `data` value returned from the hook. In other words:

```ts
function MyComponent() {
  const [execute] = useLazyQuery(QUERY);

  function handleClick() {
    const { data } = await execute();
  }
}
```

Notice how the function is the only value destructured from `useLazyQuery`. This caused component rerenders for no reason since you're only consuming the value where the interaction was performed. `client.query` is much better in these situations since you get the same return value, but without the rerender penalty. We added this to our [docs](https://www.apollographql.com/docs/react/data/queries#using-the-promise-returned-from-the-execution-function) (see the tip underneath that code block).

This might not affect you as much on the angular side, but it was important enough to keep at least for React users for this reason.

I would agree though that `watchQuery` should be what you use more often than not 🙂
